src/abstractdataframe/join.jl

```diff
@@ -81,16 +81,349 @@ end
 
 Base.length(x::RowIndexMap) = length(x.orig)
 
-# composes the joined data table using the maps between the left and right
-# table rows and the indices of rows in the result
-
 _rename_cols(old_names::AbstractVector{Symbol},
              renamecols::Union{Function, Symbol, AbstractString},
              exclude::AbstractVector{Symbol} = Symbol[]) =
     Symbol[n in exclude ? n :
            (renamecols isa Function ? Symbol(renamecols(string(n))) : Symbol(n, renamecols))
            for n in old_names]
 
+prepare_on_col() = throw(ArgumentError("at least one on column required when joining"))
+prepare_on_col(c::AbstractVector) = c
+prepare_on_col(cs::AbstractVector...) = tuple.(cs...)
+
+# Return if it is allowed to use refpool instead of the original array for joining.
+# There are multiple conditions that must be met to allow for this.
+# If it is allowed we are sure that nothing can be used as a sentinel
+function check_mapping_allowed(short::AbstractVector,
+                               refarray_long::AbstractVector,
+                               refpool_long, invrefpool_long)
+    if isempty(short) ||
+       isnothing(refpool_long) ||
+       isnothing(invrefpool_long) ||
+       eltype(refarray_long) isa Union{Signed, Unsigned}
+        return false
+    else
+        return true
+    end
+end
```

Comment:
AFAICT the `eltype` check is backwards. Also this can be simplified:
```suggestion
check_mapping_allowed(short::AbstractVector, refarray_long::AbstractVector,
                      refpool_long, invrefpool_long) =
    !isempty(short) && !isnothing(refpool_long) && !isnothing(invrefpool_long) &&
        eltype(refarray_long) isa Union{Signed, Unsigned}
```